ccmlib/common.py

```diff
@@ -731,9 +742,118 @@ def _get_jdk_version(version):
     return re.search(ver_pattern, str(version)).groups()[0] + ".0"
 
 
+def update_java_version(jvm_version=None, install_dir=None, cassandra_version=None, env=None,
+                        for_build=False, info_message=None):
+    """
+    Updates or fixes the Java version (JAVA_HOME environment).
+    If 'jvm_version' is explicitly set, that one will be used.
+    Otherwise, the Java version will be guessed from the provided 'cassandra_version' parameter.
+    If the version-parameters are not specified, those will be inquired from the 'install_dir'.
+    :param jvm_version: The Java version to use - must be the major Java version number like 8 or 11.
+    :param install_dir: Software installation directory.
+    :param cassandra_version: The Cassandra version to consider for choosing the correct Java version.
+    :param env: Current OS environment variables.
+    :param for_build: whether the code should check for a valid version to build or run bdp. Currently only
+    applies to source tree that have a 'build-env.yaml' file.
+    :param info_message: String logged with info/error messages
+    :return: the maybe updated OS environment variables. If 'env' was 'None' and JAVA_HOME needs to be set,
+    the result will also contain the current OS environment variables from 'os.environ'.
+    """
+
+    env = env if env else os.environ
+
+    current_java_version = get_jdk_version_int()
+    current_java_home_version = get_jdk_version_int('{}/bin/java'.format(env['JAVA_HOME'])) if 'JAVA_HOME' in env else current_java_version
+    # Code to ensure that we start C* using the correct Java version.
+    # This is important especially after DB-468 (Java 11 support) when dtests are run using Java 11
+    # but a "manually" configured (set_install_dir()) different version requires Java 8.
+    return _update_java_version(current_java_version, current_java_home_version,
+                                jvm_version=jvm_version, install_dir=install_dir,
+                                cassandra_version=cassandra_version, env=env,
+                                for_build=for_build, info_message=info_message)
+
+
+def _update_java_version(current_java_version, current_java_home_version,
+                         jvm_version=None, install_dir=None, cassandra_version=None, env=None,
+                         for_build=False, info_message=None):
+    # Internal variant accessible for tests
+
+    if env is None:
+        raise RuntimeError("env passed to _update_java_version must not be None")
+
+    if cassandra_version is None and install_dir:
+        cassandra_version = get_version_from_build(install_dir)
+
+    # conservative Java version defaults
+    build_versions = [8]
+    run_versions = [8]
+
+    if not jvm_version:
+        if cassandra_version >= '4.0':
+            build_versions = [8, 11]
+            run_versions = [8, 11, 12, 13, 14, 15, 16, 17]
+        # Cassandra versions 3.x use the defaults
+        elif cassandra_version < '3.0':
```

Comment:
seems like we don't need that `else` branch, since the values are the same as the defaults?